server/src/remote.rs

```diff
@@ -296,19 +304,236 @@ impl FlightService for FlightServiceHandler {
+
+    info!("Created table '{}'.", table_name);
+    catalog.table_metadata.push(new_table_metadata);
+
+    Ok(())
+}
+
+/// Create a model table, add it to the metadata database tables, and add it to the catalog.
+/// If the table already exists or if the model table cannot be created, return [`Status`] error.
+fn create_model_table(
+    mut catalog: RwLockWriteGuard<Catalog>,
+    table_name: String,
+    schema: Schema,
+    tag_column_indices: Vec<u8>,
+    timestamp_column_index: u8,
+) -> Result<(), Status> {
+    // If the table already exists, return an error.
+    let mut existing_tables = catalog.new_model_table_metadata.iter();
+    if existing_tables.any(|table| table.name == table_name) {
+        let message = format!("Model table with name '{}' already exists.", table_name);
+        return Err(Status::already_exists(message));
+    }
+
+    let model_table_metadata = NewModelTableMetadata::try_new(
+        table_name.clone(),
+        schema,
+        tag_column_indices,
+        timestamp_column_index,
+    )
+    .map_err(|error| Status::invalid_argument(error.to_string()))?;
+
+    // Persist the new model table to the metadata database.
+    let database_path = catalog.data_folder_path.join("metadata.sqlite3");
+    save_model_table_to_database(database_path, &model_table_metadata)
+        .map_err(|error| Status::internal(error.to_string()))?;
+
+    info!("Created model table '{}'.", table_name);
+    catalog.new_model_table_metadata.push(model_table_metadata);
+
+    Ok(())
+}
+
+/// Save the created model table to the database. This includes creating a tags table for the
+/// model table, adding a new row in the model_table_metadata table, and adding a row to the
+/// model_table_columns table for each field column.
+fn save_model_table_to_database(
+    database_path: PathBuf,
+    model_table_metadata: &NewModelTableMetadata,
+) -> Result<(), rusqlite::Error> {
+    // Create a transaction to ensure the database state is consistent across tables.
+    let mut connection = Connection::open(database_path)?;
+    let transaction = connection.transaction()?;
+
+    // Add a column definition for each tag field in the schema.
+    let mut tag_columns = "".to_owned();
+    let mut tag_indices = model_table_metadata.tag_column_indices.iter().peekable();
+
+    while let Some(index) = tag_indices.next() {
+        // If it is the last column in the query, adding a "," results in an SQL syntax error.
+        let suffix = if tag_indices.peek().is_none() {""} else {", "};
+        let field = model_table_metadata.schema.field(*index as usize);
+
+        tag_columns.push_str(format!("{} TEXT NOT NULL{}", field.name(), suffix).as_str());
+    }
+
+    // Create a table_name_tags SQLite table to save the 54-bit tag hashes when ingesting data.
+    // The query is executed with a formatted string since CREATE TABLE cannot take parameters.
+    transaction.execute(format!("CREATE TABLE {}_tags (hash BLOB PRIMARY KEY, {})",
```

Comment:
It is unclear why the 54-bits are stored as a BLOB instead of a 64-bit BIGINT (although using BIGINT is mostly for documentation as [SQLite does not differentiate between INTEGER and BIGINT](https://www.sqlite.org/datatype3.html) as far as I know). By storing the 54-bits as a 64-bit BIGINT with zero bits used for the 10 least significant bits, it should be possible to directly combined it with the column index through a bitwise OR.

Reply:
I have changed it to `BIGINT`.